### ISSUE TITLE
NanoGPT training crashes on ROCm

### ISSUE THREAD

**Issue by jglaser** [pytorch/pytorch#95808]
### 🐛 Describe the bug

When training nanoGPT on ROCm (MI250x) GPUs, I get a memory access fault:

```
$ python train.py config/train_shakespeare_char.py --compile=False
Overriding config with config/train_shakespeare_char.py:
# train a miniature character-level shakespeare model
# good for debugging and playing on macbooks and such

out_dir = 'out-shakespeare-char'
eval_interval = 250 # keep frequent because we'll overfit
eval_iters = 200
log_interval = 10 # don't print too too often

# we expect to overfit on this small dataset, so only save when val improves
always_save_checkpoint = False

wandb_log = False # override via command line if you like
wandb_project = 'shakespeare-char'
wandb_run_name = 'mini-gpt'

dataset = 'shakespeare_char'
batch_size = 64
block_size = 256 # context of up to 256 previous characters

# baby GPT model :)
n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3 # with baby networks can afford to go a bit higher
max_iters = 5000
lr_decay_iters = 5000 # make equal to max_iters usually
min_lr = 1e-4 # learning_rate / 10 usually
beta2 = 0.99 # make a bit bigger because number of tokens per iter is small

warmup_iters = 100 # not super necessary potentially

# on macbook also add
# device = 'cpu'  # run on cpu only
# compile = False # do not torch compile the model

Overriding: compile = False
found vocab_size = 65 (inside data/shakespeare_char/meta.pkl)
Initializing a new model from scratch
WARNING: using slow attention. Flash Attention atm needs PyTorch nightly and dropout=0.0
WARNING: using slow attention. Flash Attention atm needs PyTorch nightly and dropout=0.0
WARNING: using slow attention. Flash Attention atm needs PyTorch nightly and dropout=0.0
WARNING: using slow attention. Flash Attention atm needs PyTorch nightly and dropout=0.0
WARNING: using slow attention. Flash Attention atm needs PyTorch nightly and dropout=0.0
WARNING: using slow attention. Flash Attention atm needs PyTorch nightly and dropout=0.0
number of parameters: 10.65M
using fused AdamW: True
step 0: train loss 4.2874, val loss 4.2823
Memory access fault by GPU node-4 (Agent handle: 0x6f9b500) on address (nil). Reason: Unknown.
Aborted
```

### Expected outcome
Training continues

### Versions

Please see the [nanoGPT repository](https://github.com/karpathy/nanoGPT) for how to install the reproducer and dependencies.

In a nutshell:
```
git clone https://github.com/karpathy/nanoGPT.git
cd nanoGPT
python data/shakespeare_char/prepare.py
python train.py config/train_shakespeare_char.py --compile=False
```

Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: SUSE Linux Enterprise Server 15 SP3 (x86_64)
GCC version: (SUSE Linux) 7.5.0
Clang version: 14.0.2  (ecfd9ef4dfd5696cd449133c0da0293d503c2f21)
CMake version: version 3.17.0
Libc version: glibc-2.31

Python version: 3.8.16 (default, Jan 17 2023, 23:13:24)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.3.18-150300.59.87_11.0.78-cray_shasta_c-x86_64-with-glibc2.17
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          128
On-line CPU(s) list:             0-127
Thread(s) per core:              2
Core(s) per socket:              64
Socket(s):                       1
NUMA node(s):                    4
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           48
Model name:                      AMD EPYC 7A53 64-Core Processor
Stepping:                        1
Frequency boost:                 enabled
CPU MHz:                         3093.761
CPU max MHz:                     2000.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        3992.40
Virtualization:                  AMD-V
L1d cache:                       2 MiB
L1i cache:                       2 MiB
L2 cache:                        32 MiB
L3 cache:                        256 MiB
NUMA node0 CPU(s):               0-15,64-79
NUMA node1 CPU(s):               16-31,80-95
NUMA node2 CPU(s):               32-47,96-111
NUMA node3 CPU(s):               48-63,112-127
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==1.8.6
[pip3] torch==2.0.0a0+git10a1efb
[pip3] torchmetrics==0.10.3
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-lightning         1.8.6                    pypi_0    pypi
[conda] torch                     2.0.0a0+git10a1efb          pypi_0    pypi
[conda] torchmetrics              0.10.3                   pypi_0    pypi

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport

**Comment by sunway513** [comment]
Hi @jglaser , which docker container did you use?
I pulled the docker container `rocm/pytorch:latest`, and I'm able to use the following steps for training:
```
git clone https://github.com/karpathy/nanoGPT.git
cd nanoGPT
python data/shakespeare_char/prepare.py
python train.py config/train_shakespeare_char.py --compile=False
```
<img width="494" alt="image" src="https://user-images.githubusercontent.com/1049239/222287091-348906b9-86f5-4da0-812b-e345834bf6ea.png">

**Comment by jglaser** [comment]
I am using a locally built pyorch, installed via pip. I did not try the docker container.

**Comment by jglaser** [comment]
P.S.: Source version

```
$ cd pytorch
$ git describe
ciflow/periodic/79617-8852-gd37748adad
```

**Comment by jglaser** [comment]
> Hi @jglaser , which docker container did you use? I pulled the docker container `rocm/pytorch:latest`, and I'm able to use the following steps for training:
> 
> ```
> git clone https://github.com/karpathy/nanoGPT.git
> cd nanoGPT
> python data/shakespeare_char/prepare.py
> python train.py config/train_shakespeare_char.py --compile=False
> ```

I will also try your container

**Comment by jglaser** [comment]
I spoke too soon. The system I am running on is an HPC system and does not currently support containers, unfortunately
https://docs.olcf.ornl.gov/systems/crusher_quick_start_guide.html

On a side note, I noticed in your output that the line "using fused AdamW: False"  differs (mine says `True`). 

Just to make sure, would it be possible for you to run the example in `rocgdb`?

**Comment by sunway513** [comment]
I was able to reproduce your reported issue with the nightly WHL, we'll take a deeper look there and update.. thanks for reporting the problem. 

Also I can confirm the workload trains well with the stable release whl installed with the following instruction:
```
pip3 install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/rocm5.2
```
Or use the `rocm/pytorch:latest` container..

**Comment by jglaser** [comment]
Thanks.  Please note that I provided a fix in this PR #95810
